lib/oidc/key_service.rb

```diff
@@ -7,23 +7,45 @@ class KeyService
     @mutex = Mutex.new
     # Map from kid to OpenSSL::RSA::PKey for all current keys
     @current_keys = {}
-
-    class << self
-      attr_reader :current_keys
-    end
+    @cache_miss_kids = {}
+    KID_CACHE_PERIOD = 60
+    KID_CACHE_MAX_SIZE = 1000
```

Comment:
The cache period and max size are both pretty arbitrary. The max size could probably be larger without much risk, but given that this is per instance, per deploy, I think 1000 should be fine for now.

Reply:
Specifying the unit (`KID_CACHE_PERIOD_SECS`) for the period could be a nice readability improvement in the comparison logic below.

I would suggest settings the max cache size lower to start. A malicious user could easily generate enough requests to fill the cache whether it's 100 entries, 1,000 entries, or 10,000 entries. A smaller cache size minimizes the time to update it, leaving more compute time and memory for other activities. Due to the ineffectiveness of memory compaction in ruby, the added memory fragmentation caused by the cache is more of a concern than the absolute memory usage at any given instant.